### ISSUE TITLE
TPC requires server-to-server XRootDTransport authentication

### ISSUE THREAD

**Issue by olifre** [xrootd/xrootd#694]
As documented http://xrootd.org/doc/dev4/tpc_protocol.htm , it seems no separate authentication between the servers is needed. 

However, this seems to be actually required. Consider the following setup:
- Source: Server A with sec.protocol gsi
- Destination: Server B with sec.protocol gsi
- Client able to authenticate to both

Then, I observe the following when looking at the debug logs of the `xrdcp` started by Server B to pull from Server A:
```
[Debug  ][TaskMgr           ] Registering task: "TickGeneratorTask for: ServerA.example.com:1094" to be run at: [2018-04-25 00:03:37 +0200]
[Dump   ][PostMaster        ] [ServerA.example.com:1094 #0] Sending message kXR_open (file: /some/source/file?tpc.key=XXXXXX&tpc.org=XXXXXX, mode: 00, flags: kXR_open_read kXR_async kXR_retstat ) (0xb1bfc0) through substream 0 expecting answer at 0
[Debug  ][PostMaster        ] [ServerA.example.com:1094] Found 1 address(es): [::ffff:SOMEIP]:1094
[Debug  ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Attempting connection to [::ffff:SOMEIP]:1094
[Debug  ][Poller            ] Adding socket 0xb20710 to the poller
[Debug  ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Async connection call returned
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Sending out the initial hand shake + kXR_protocol
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Wrote a message:  (0xe0000950), 44 bytes
[Dump   ][XRootDTransport   ] [msg: 0xe0000950] Expecting 8 bytes of message body
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received message header, size: 8
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received a message of 16 bytes
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Got the server hand shake response (type: server [], protocol version 310)
[Dump   ][XRootDTransport   ] [msg: 0xe0000950] Expecting 8 bytes of message body
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received message header, size: 8
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received a message of 16 bytes
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] kXR_protocol successful (type: server [], protocol version 310)
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Sending out kXR_login request, username: xrootd, cgi: ?xrd.cc=de&xrd.tz=1&xrd.appname=xrdcp.orig&xrd.info=&xrd.hostname=ServerB.example.com&xrd.rn=v4.8.2, dual-stack: false, private IPv4: true, private IPv6: false
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Wrote a message:  (0xe0000bd0), 133 bytes
[Dump   ][XRootDTransport   ] [msg: 0xe0000950] Expecting 61 bytes of message body
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received message header, size: 8
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received a message of 69 bytes
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Logged in, session: 02000000591900001800000082000000
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Authentication is required: &P=gsi,v:10300,c:ssl,ca:XXXXXX.0|XXXXXX.0
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Sending authentication data
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Trying to authenticate using gsi
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Cannot get credentials for protocol gsi: Secgsi: ErrParseBuffer: error getting user proxies: kXGS_init
[Error  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] No protocols left to try
[Error  ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Socket error while handshaking: [FATAL] Auth failed
[Debug  ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Closing the socket
```
Finally, I see `[FATAL] Auth failed.` and TPC does not take place. 

However, it works just fine if the Source-server has the unsafe "unix" authentication enabled on ServerA:
```
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Wrote a message:  (0xc000bd0), 133 bytes
[Dump   ][XRootDTransport   ] [msg: 0xc000950] Expecting 68 bytes of message body
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received message header, size: 8
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received a message of 76 bytes
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Logged in, session: 020000004d5b00001800000003000000
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Authentication is required: &P=gsi,v:10300,c:ssl,ca:XXXXXX.0|XXXXXX.0&P=unix
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Sending authentication data
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Trying to authenticate using gsi
[Dump   ][TaskMgr           ] Running task: "FileTimer task"
[Dump   ][TaskMgr           ] Will rerun task "FileTimer task" at [2018-04-25 00:35:04 +0200]
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Cannot get credentials for protocol gsi: Secgsi: ErrParseBuffer: error getting user proxies: kXGS_init
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Trying to authenticate using unix
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Wrote a message:  (0xc0faa50), 43 bytes
[Dump   ][XRootDTransport   ] [msg: 0xc0faa50] Expecting 0 bytes of message body
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received message header, size: 8
[Dump   ][AsyncSock         ] [ServerA.example.com:1094 #0.0] Received a message of 8 bytes
[Debug  ][XRootDTransport   ] [ServerA.example.com:1094 #0.0] Authenticated with unix.
[Debug  ][PostMaster        ] [ServerA.example.com:1094 #0] Stream 0 connected.
```
It seems login happens as user `xrootd`, even though this user is not listed in `acc.authdb`. 

Now, there are two issues here:
- TPC does not seem to work unless the security protocol settings allow authentication of ServerB to ServerA. 
- I don't think this security is needed, since the `tpc.key` is already shared with both servers at this point. It's also not obvious from http://xrootd.org/doc/dev4/tpc_protocol.htm that this is to be expected. 
- In the common Grid DDM usecase with GSI authentication, cross-server authentication is not  really foreseen.... 
- The only "easy" hack appears to be to enable user authentication. It seems some endpoints, e.g. eosatlas, are doing that... 

Is there another configuration "trick" one could use to have TPC working in a safe configuration?

**Comment by abh3** [comment]
The TPC protocol is independent of any particular authentication. That is 
why you can use any authentication you wish -- even different methods for 
the source and destination. Being independent of the authentication method 
being used does *not* mean it does not require authentication. All sites 
typically configure their servers to requires some kind of authentication 
irrespective of what is being sent on the wire (e.g. TPC). If that is the 
case, then clearly, server to server cmmunications need to authenticate as 
well. The TPC protocol spec merely specifies the exchanges *after* any 
required authentication occurs.

Andy

**Comment by olifre** [comment]
> All sites typically configure their servers to requires some kind of authentication irrespective of what is being sent on the wire (e.g. TPC). If that is the case, then clearly, server to server cmmunications need to authenticate as well. 

Understood - is there a way to specify within the xrootd configuration, that *only* for TPC being on the wire, a different `sec.protocol` is being applied? I don't find that in the documentation. 
If there is no way yet, would this be a useful new feature?

Otherwise, I don't see how Grid usecases can be covered securely as things are now. The straightforward solution would be to enable "unix", which is not save - a secure way would be to give all servers robot certificates in addition to their hostcerts. Of course, the final path to use is not a problem of xrootd, but rather something DDM should figure out and solve.

**Comment by alrossi** [comment]
Hi, I am running up against the same issue as Oliver.  I am trying to enable TPC for the dCache xroot door.  I have an unauthenticated version working, but now need to handle authenticated.  In order to test this, I set up two xrootd servers with gsi and default authorization (Authfile).   I see the same failure (after the client sends the second SYNC to the destination).   Is the story that we just need to enable user authentication as well, or is there some other best practice which should be observed?  Thanks,  Al

**Comment by olifre** [comment]
@alrossi I asked on the DDM list to try to learn about a best-practice - I think this xrootd bugtracker is the wrong place, since it's only for development of xrootd, not the application in WLCG. 
However, what I learned was not really helpful: I did not get indications of any best practice... 
Finally, I decided to request a robot / service certificate for our xrootd servers from GridKa, register it with VOMS, and have our servers use it. 
Like this, our servers can authenticate with GSI and TPC will work fine. 

Other sites, in case they want to fetch data from us, naturally also need to authenticate with GSI, since we did not enable unsafe user auth. I asked on the DDM list whether this is a problem, but was not made aware that this could be an issue. 

In any case, another issue was that the ddmadmin-user of FTS did not authenticate via GSI, which also made transfers fail. I whitelisted the corresponding certificate explicitly in the Authfile for now and told the DDM list about that - again, no response yet. 

In summary, since service / robot certificates are also needed for xcache, I would expect this is the way to go. EOSATLAS, on the other hand, has user auth enabled. I asked them, and the reply I got was "that's expected, we need that for TPC". I asked back why they don't use a robot cert - no response.

**Comment by alrossi** [comment]
@olifre Thank you for this information, Oliver.  It will at least help me determine my next action.  As I am not so much concerned with system policy as development, I just needed a way to (a) make sure the servers are behaving correctly and then (b) make sure I do the same thing in dCache.  Yes, I was reluctant to post this here, unsure as to whether this was a missing feature or not.

**Comment by olifre** [comment]
@alrossi Ok! I think what could be interesting for dCache is to have a similar way to specify the TPC command, for example to replace it with a script. 
It might be that xrootd at some point learns to do VOMS authentication by itself, but right now, I have to solve this in the external script (which also fixes permissions etc.) - and I guess the same will be true for many dCache users. 
Probably, you should also wait for a reply from @abh3 to see whether anything else is in the queue to change the authentication - if so, he will know ;-).

**Comment by wyang007** [comment]
I think some kind of authentication based on shared secret/tpc.key is still useful. But perhaps until we have SSL enabled in xroot protocol, this is not quite secure? Another way is to use proxy delegation. This works if everyone uses GSI (which is true in LHC, for now). But if site A uses GSI and site B use ABC, then delegation doesn’t work.

For now the only thing that works is to provide this robot certificate… Why ddmadmin used by FTS does not work is strange. Does the proxy has VOMS attribute?


--
Wei Yang  |  mailto:yangw@slac.stanford.edu  |  650-926-3338 (O)

**Comment by olifre** [comment]
> I think some kind of authentication based on shared secret/tpc.key is still useful.

As far as I understand, currently, when one server starts xrdcp to start up a TPC transfer, **authentication** happens before the TPC key is transferred to the other end. Hence, this needs to happen between the two servers, either via shared secret, GSI or something else. Only after that, **authorization** via the TPC key can step in. 



> Why ddmadmin used by FTS does not work is strange. Does the proxy has VOMS attribute?

No, the ddmadmin robot does not transfer any VOMS attrobutes, and that's exactly the problem. I only see the GSI certificate, but no VO or role.

**Comment by alrossi** [comment]
Is proxy delegation enabled for TPC?  How do you turn it one?  I've been trying to find this in the documentation.

**Comment by olifre** [comment]
> Is proxy delegation enabled for TPC?

I think it's not implemented yet - or is it?

**Comment by alrossi** [comment]
I get the impression not.  It's in the list from 2016, but I don't see anything in the release notes.

**Comment by abh3** [comment]
SSI should be as secure as any other shared secret protocol. I think what people don't like is that the proxy server's cert is used to access the file, which actually is not true. The proxy cert is merely used to a) make sure the server is allowed to access files, and b) the particular file being accessed was also accessible by the person requesting the copy. Tied into the proxy cert is the annoyance of getting another cert in the first place. I suppose even robocerts aren't easy enough to deal with. In tat respect, we are working to enable client cert delegation. But that isn't necessarily a long-lived solution since, as Wei points out, in the future x509 might not be the universal authentication protocol.

**Comment by alrossi** [comment]
Thank you all for the clarifications.  I will adopt the common solution (robot cert) for my test setup.

**Comment by bbockelm** [comment]
Just as a note as a point of comparison - the HTTPS implementation in Xrootd allows clients that perform HTTPS to be authenticated or unauthenticated, adjusting the session identity accordingly.

Why can't we do this here?  Simply establish a GSI connection but consider the client side unauthenticated?

**Comment by olifre** [comment]
> Simply establish a GSI connection but consider the client side unauthenticated?

I'm not sure I understand the suggestion - this ticket has been about xrdcp executed on a server after being told by a client to perform a third-party-copy. 
If the server has a robot cert, everything already works fine. My initial question was if things can work without a robot cert (and then, naturally, GSI can not be used), since the tpc key is available. However, the tpc key only grants authorization after an initial authentication. 
The two solutions to get servers authenticated to other servers using GSI are:
- Get a robot cert, so the server can do GSI. That works fine. 
- Implement proxy delegation. This is under way. 

What exactly is your proposal? 
To allow unauthorized clients to proceed to the authorization stage? 
My expectation is that is unsafe, since also unix auth is regarded as unsafe. 

On a side note, xrootd's HTTP implementation is still not suitable for WLCG, e.g. due to https://github.com/xrootd/xrootd/issues/691 , so it's not useful for WLCG replication anyways right now.

**Comment by bbockelm** [comment]
Why precisely does GSI authentication require a client certificate?  There's no underlying technical reason - it's just that the current implementation requires it.

Indeed - unauthenticated clients should proceed to the authorization stage just like they do elsewhere.  If they are unauthenticated and have an authorized token for a TPC, then they should indeed be authorized.  That's the whole point of a bearer token.

FWIW - FTS-based transfers work regardless of #691 and we are indeed using them for WLCG replication.  I would hardly say that the presence of a technical bug (which should be followed up and fixed, of course) in the handling of a Nagios test makes any implications about "suitability".  It just means there's continued room for improvement.

**Comment by olifre** [comment]
Dear @bbockelm ,

thanks for the explanations! 

> If they are unauthenticated and have an authorized token for a TPC, then they should indeed be authorized. That's the whole point of a bearer token.

Ok - this is exactly what my initial question in this ticket was, but better formulated. 
Let's see what @abh3 comments on this. 

Also, thanks for the WebDAV explanation. Since I have never seen transfers using WebDAV from / to our site, and only the failing tests, and we currently have an issue with xroot (almost *all* other sites apart from EOSATLAS are not TPC ready, so they can't deliver data to us) and DDM people did not offer WebDAV as solution, I will take this point to ask them why WebDAV is not used. Thanks!

**Comment by alrossi** [comment]
Brian,


These are also our (Dmitry, Al) thoughts here at FNAL.


(I am currently engaged in building into dCache native xrootd TPC).


Just to weigh in on the question ...

________________________________________________
Albert L. Rossi
Application Developer & Systems Analyst III
Scientific Computing Division, Data Movement Development
FCC 229A
Mail Station 369 (FCC 2W)
Fermi National Accelerator Laboratory
Batavia, IL 60510
(630) 840-3023


________________________________

**Comment by abh3** [comment]
Hello @bbockelm @olifre 
Well, this is actually more involved than one might think. Technically, if you have a bearer token you don't need to authenticate. On the other hand, requiring authentication gives you a chance, perhaps small but in our case not very small, to detect stolen bearer tokens (e.g. a CMS user  stealing an ATLAS file :-)

I suspect many experiments will want authenticated third party transfers. That said, there is technically no reason why http or xrootd can't support unauthenticated transfers as long as the bearer token is valid.

**Comment by simonmichal** [comment]
Since we implemented TPC with delegation (and released in 4.9.0) I suppose we can close this one?

**Comment by olifre** [comment]
Oh yes, of course! 
I'll close it with this comment, thanks for the reminder!
